### PR TITLE
Various enhancements

### PR DESCRIPTION
**Pull request by kpouget**
Sharing the enhancements I had to do on `benchmark-runner` to get my workload running. To be discussed.

-* `template_operations: do not crash if 'shared' is missing in the template`

---

* `environment_variables: read settings from an environment file`

---

* `workloads_operations: make 'oc' a class singleton`

---

* `benchmark_runner/common/ssh/ssh_exceptions.py: fix typo`

---

* `benchmark_runner/common/oc/oc.py: _create_async: raise an excption if the command fails`

---

* `benchmark_runner/main/environment_variables.py: allow overriding the namespace with env var`

---

* `benchmark_runner/main: allow setting the RUNNER_TYPE with an env variable`

---

* `benchmark_runner: allow reading the config from the CLI args`

---

* `benchmark_runner/common/template_operations/template_operations.py: allow storing the template data in the workload directory`

---

* `benchmark_runner: allow overriding initialize/finalize_workload methods`

---

* `benchmark_runner/main/environment_variables.py: allow setting RUN_ARTIFACTS_PATH`

---

* `benchmark_runner: report benchmark failure`

---

* `benchmark_runner/workloads/workloads_operations.py: remove trailing spaces`

---

### CODE REVIEW

**Comment by kpouget** [comment]
```
benchmark_runner/common/template_operations/template_operations.py:97: in generate_yamls_internal
    if self._environment_variables_dict['template_in_workload_dir'] == "True":
E   AttributeError: 'TemplateOperations' object has no attribute '_environment_variables_dict'
```
I changed that this morning to avoid relying on `mpi` in the name, I'll fix that

**Comment by RobertKrawitz** [comment]
@kpouget you also need to update the golden files; the instructions are in the HOWTO.

File: benchmark_runner/common/elasticsearch/elasticsearch_operations.py [hunk 1]
```diff
@@ -28,9 +28,19 @@ def __init__(self, es_host: str, es_port: str, es_user: str = '', es_password: s
...
+        ssl_context.check_hostname = False
+        ssl_context.verify_mode = ssl.CERT_NONE
+        import urllib3
+        urllib3.disable_warnings()
+
         if self.__es_password:
-            self.__es_url = f"http://{self.__es_user}:{self.__es_password}@{self.__es_host}:{self.__es_port}"
-            self.__es = Elasticsearch([self.__es_url])
+            self.__es_url = f"https://{self.__es_user}:{self.__es_password}@{self.__es_host}:{self.__es_port}"
+            self.__es = Elasticsearch([self.__es_url], ssl_context=ssl_context)
```
Comment (ebattat) [inline benchmark_runner/common/elasticsearch/elasticsearch_operations.py:43]: Need to keep password option seperatly and add more option for https in environment variable:
https://github.com/redhat-performance/benchmark-runner/blob/main/benchmark_runner/main/environment_variables.py#L39

File: benchmark_runner/common/ssh/ssh.py [hunk 2]
```diff
@@ -22,6 +22,7 @@ def run(self, cmd: str, is_check: bool = False, background: bool = False):
         :return:
         """
         try:
+            print("|", cmd)
```
Comment (ebattat) [inline benchmark_runner/common/ssh/ssh.py:25]: please replace it to logger.info.
Do we really need it ?

File: benchmark_runner/common/ssh/ssh.py [hunk 3]
```diff
@@ -35,7 +36,9 @@ def run(self, cmd: str, is_check: bool = False, background: bool = False):
                 output = subprocess.getoutput(cmd)
             return output
         except subprocess.CalledProcessError as err:
-            logger.error(f'subprocess Status : FAIL: {err.returncode} {err.output}')
-            raise SSHSubprocessError()
+            logger.info(f'subprocess Command: {cmd}')
```
Comment (ebattat) [inline benchmark_runner/common/ssh/ssh.py:39]: Why did you delete the logger.error ?

File: benchmark_runner/common/template_operations/template_operations.py [hunk 4]
```diff
@@ -118,9 +133,9 @@ def get_sub_dict(dictionary: dict, key: str, value: str):
             else:
                 file_components = os.path.splitext(filename)
                 template_file = f'{file_components[0]}_template{file_components[1]}'
-            with open(os.path.join(workload_dir_path, 'internal_data', template_file)) as f:
```
Comment (ebattat) [inline benchmark_runner/common/template_operations/template_operations.py:121]: Why did you delete ''internal_data' ?

File: benchmark_runner/common/template_operations/template_operations.py [hunk 5]
```diff
@@ -118,9 +133,9 @@ def get_sub_dict(dictionary: dict, key: str, value: str):
...
+            with open(os.path.join(workload_dir_path, template_file)) as f:
                 template = Template(f.read())
-            answer[filename] = template.render(render_data)
+            answer[filename] = template.render(render_data) + "\n"
```
Comment (ebattat) [inline benchmark_runner/common/template_operations/template_operations.py:138]: Why we need a newline at the end of render ?

File: benchmark_runner/main/environment_variables.py [hunk 6]
```diff
@@ -59,7 +76,9 @@ def __init__(self):
 
         # Choose default namespace
         base_workload = self._environment_variables_dict['workload'].split('_')[0]
-        if base_workload in self._environment_variables_dict['workload_namespaces']:
+        if os.environ.get('NAMESPACE'):
```
Comment (ebattat) [inline benchmark_runner/main/environment_variables.py:79]: Need to verify that we keep backward compatibility for other namespaces

File: benchmark_runner/main/environment_variables.py [hunk 7]
```diff
@@ -198,10 +228,10 @@ def __init__(self):
 
         # ElasticSearch url
         if self._environment_variables_dict.get('elasticsearch_password', ''):
-            self._environment_variables_dict['elasticsearch_url'] = f"http://{self._environment_variables_dict.get('elasticsearch_user', '')}:{self._environment_variables_dict.get('elasticsearch_password', '')}@{self._environment_variables_dict.get('elasticsearch', '')}:{self._environment_variables_dict.get('elasticsearch_port', '')}"
+            self._environment_variables_dict['elasticsearch_url'] = f"https://{self._environment_variables_dict.get('elasticsearch_user', '')}:{self._environment_variables_dict.get('elasticsearch_password', '')}@{self._environment_variables_dict.get('elasticsearch', '')}:{self._environment_variables_dict.get('elasticsearch_port', '')}"
```
Comment (ebattat) [inline benchmark_runner/main/environment_variables.py:231]: Please add separate variable for elasticsearch_ssl

File: benchmark_runner/main/environment_variables.py [hunk 8]
```diff
@@ -198,10 +228,10 @@ def __init__(self):
...
         else:
             if self._environment_variables_dict['elasticsearch'] and self._environment_variables_dict.get('elasticsearch_port', ''):
-                self._environment_variables_dict['elasticsearch_url'] = f"http://{self._environment_variables_dict.get('elasticsearch', '')}:{self._environment_variables_dict.get('elasticsearch_port', '')}"
+                self._environment_variables_dict['elasticsearch_url'] = f"https://{self._environment_variables_dict.get('elasticsearch', '')}:{self._environment_variables_dict.get('elasticsearch_port', '')}"
```
Comment (ebattat) [inline benchmark_runner/main/environment_variables.py:234]: Please add separate variable for elasticsearch_ssl

File: benchmark_runner/main/main.py [hunk 9]
```diff
@@ -151,11 +155,15 @@ def run_benchmark_runner_workload():
...
         update_ci_status()
-    elif environment_variables.get_workload_namespace(workload) == 'benchmark-operator':
+    elif is_benchmark_operator_workload:
         run_benchmark_operator_workload()
-    elif environment_variables.get_workload_namespace(workload) == 'benchmark-runner':
-        run_benchmark_runner_workload()
 
+    elif is_benchmark_runner_workload:
+        success = run_benchmark_runner_workload()
+        if not success: sys.exit(1)
```
Comment (ebattat) [inline benchmark_runner/main/main.py:163]: Please raise custom error, its not recommend to use sys.exit

File: benchmark_runner/main/main.py [hunk 10]
```diff
@@ -151,11 +155,15 @@ def run_benchmark_runner_workload():
...
+    else:
+        print("ERROR: could not determine the type of execution :/")
+        sys.exit(2)
```
Comment (ebattat) [inline benchmark_runner/main/main.py:166]: Please raise custom error, its not recommend to use sys.exit

File: benchmark_runner/workloads/workloads.py [hunk 11]
```diff
@@ -22,12 +22,33 @@ def run(self):
...
+                if cls[1]().run() == False:
+                    success = False
+
+        try:
+            _finalize_workload = getattr(workload_module, "finalize_workload")
+            finalize_workload = lambda: _finalize_workload(self)
+        except AttributeError:
+            print(f"WARNING: {workload} module has no finalize_workload method. Using the default one.")
+            finalize_workload = self.finalize_workload
+        finalize_workload()
```
Comment (ebattat) [inline benchmark_runner/workloads/workloads.py:52]: Need to verify that it keep backward compatibility

**Review by ebattat — COMMENTED** [review]
Please review my comments

File: benchmark_runner/common/oc/oc.py [hunk 12]
```diff
@@ -178,7 +178,7 @@ def _create_async(self, yaml: str):
         :return:
         """
         if os.path.isfile(yaml):
-            return self.run(f'oc create -f {yaml}')
+            return self.run(f'oc create -f {yaml}', is_check=True)
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/common/oc/oc.py:181]: > Why do we need to add is_check=True here ?

swallowing errors is bug-prone, it should be avoided whenever possible

File: benchmark_runner/common/ssh/ssh.py (hunk 3 above)
Comment (kpouget) [reply benchmark_runner/common/ssh/ssh.py:39]: I'll remove this commit, I tried to limit the flow of error messages / length of callstack to made my debugging easier, but that's personal taste

File: benchmark_runner/common/template_operations/template_operations.py (hunk 5 above)
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/template_operations.py:138]: Linux/UNIX convention is for all lines, including the last line, to be newline-terminated.

Comment (kpouget) [reply benchmark_runner/common/template_operations/template_operations.py:138]: from what I've seen, the templates files are generated without EOL at EOF,
it's best practise to have it https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: benchmark_runner/main/main.py [hunk 13]
```diff
@@ -35,18 +35,22 @@ def main():
...
+    is_benchmark_operator_workload = 'benchmark-operator' in (environment_variables.get_workload_namespace(workload), environment_variables_dict.get("runner_type"))
+    is_benchmark_runner_workload = 'benchmark-runner' in (environment_variables.get_workload_namespace(workload), environment_variables_dict.get("runner_type"))
     # workload name validation
     if workload and not ci_status:
         if workload not in environment_variables.workloads_list:
-            logger.info(f'Enter valid workload name {environment_variables.workloads_list}')
-            raise Exception(f'Not valid workload name: {workload} \n, choose one from the list: {environment_variables.workloads_list}')
+            #logger.info(f'Enter valid workload name {environment_variables.workloads_list}')
+            #raise Exception(f'Not valid workload name: {workload} \n, choose one from the list: {environment_variables.workloads_list}')
+            pass # ignore this error
```
Comment (RobertKrawitz) [inline benchmark_runner/main/main.py:46]: Why should this be ignored?

File: benchmark_runner/main/main.py [hunk 14]
```diff
@@ -151,11 +155,15 @@ def run_benchmark_runner_workload():
...
+    elif is_benchmark_runner_workload:
+        success = run_benchmark_runner_workload()
+        if not success: sys.exit(1)
+    else:
+        print("ERROR: could not determine the type of execution :/")
+        sys.exit(2)
 
-main()
 
+main()
```
Comment (RobertKrawitz) [inline benchmark_runner/main/main.py:169]: Unnecessary change

**Review by RobertKrawitz — COMMENTED** [review]
Mostly fairly small things.

File: benchmark_runner/main/main.py (hunk 9 above)
Comment (kpouget) [reply benchmark_runner/main/main.py:163]: this is the `main` function, and it's not called with `sys.exit(main())` (otherwise, I would have replaced `sys.exit(n))`  by `return n`)
it's not recommended to throw exception out of the `main()` function, it should rather return a proper error code

what about this then?
```
if not success: return 1
```
and
```
sys.exit(main())
```

File: benchmark_runner/common/ssh/ssh.py (hunk 2 above)
Comment (kpouget) [reply benchmark_runner/common/ssh/ssh.py:25]: same as below,

> I'll remove this commit, I tried to limit the flow of error messages / length of callstack to made my debugging easier, but that's personal taste

File: benchmark_runner/main/main.py (hunk 13 above)
Comment (kpouget) [reply benchmark_runner/main/main.py:46]: hum, that's something unfinished, but my overall idea is to avoid having anything related to (new) workloads outside of the workload directory.

see this commit to see how I add my workload: https://github.com/kpouget/benchmark-runner/commit/0db6a70e2e3401112a90744d20ff6db256b4a937

I think this `workloads_list` is just a safe guard, maybe avoiding running "nothing" if the workload variable isn't properly configured

I could add another env var `ignore_workloads_list` to preserve the existing behavior

File: benchmark_runner/main/environment_variables.py (hunk 6 above)
Comment (RobertKrawitz) [reply benchmark_runner/main/environment_variables.py:79]: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

File: benchmark_runner/main/main.py (hunk 13 above)
Comment (RobertKrawitz) [reply benchmark_runner/main/main.py:46]: I think for now we should not ignore the error, but please open an feature request about avoiding anything related to new workloads outside of the workload directory.  I agree with your reasoning.

**Comment by RobertKrawitz** [comment]
Looks like you need to update the golden files.

File: benchmark_runner/common/template_operations/template_operations.py [hunk 15]
```diff
@@ -67,6 +68,7 @@ def get_sub_dict(dictionary: dict, key: str, value: str):
                 return dict()
             template_data = template_root_data.get('template_data')
             shared_data = template_data.get('shared')
+            if shared_data is None: shared_data = {}
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/common/template_operations/template_operations.py:71]: fixed on next push

File: benchmark_runner/common/elasticsearch/elasticsearch_operations.py [hunk 16]
```diff
@@ -28,9 +28,19 @@ def __init__(self, es_host: str, es_port: str, es_user: str = '', es_password: s
...
+
+        import ssl
+        from elasticsearch.connection import create_ssl_context
+
+        ssl_context = create_ssl_context()
+        ssl_context.check_hostname = False
+        ssl_context.verify_mode = ssl.CERT_NONE
+        import urllib3
+        urllib3.disable_warnings()
+
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/common/elasticsearch/elasticsearch_operations.py]: done, moved this commit to a dedicated PR (https://github.com/redhat-performance/benchmark-runner/pull/394)

File: benchmark_runner/common/template_operations/template_operations.py [hunk 17]
```diff
@@ -89,8 +91,16 @@ def get_sub_dict(dictionary: dict, key: str, value: str):
             self.__standard_output_file = f"{'_'.join([self.__workload_name, self.__workload_template_kind, self.__workload_extra_name])}.yaml"
         else:
             self.__standard_output_file = f"{'_'.join([self.__workload_name, self.__workload_template_kind])}.yaml"
-        self.__standard_template_file = f"{'_'.join([self.__workload_name, self.__workload_template_kind])}_template.yaml"
-        workload_dir_path = os.path.join(self.__dir_path, self.__workload_name)
+
+        if self.__environment_variables_dict['template_in_workload_dir'] == "True":
+            workload_dir_path = os.path.join((os.path.dirname(__file__)), "..", "..", "workloads", self.__workload, "template")
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/common/template_operations/template_operations.py:97]: fix with
> workload_dir_path = os.path.join((os.path.dirname(benchmark_runner.__file__)), "workloads", self.__workload, "template")

File: benchmark_runner/common/template_operations/template_operations.py [hunk 18]
```diff
@@ -89,8 +91,16 @@ def get_sub_dict(dictionary: dict, key: str, value: str):
...
+            data_filename = 'config.yaml'
+            self.__standard_template_file = f"template.yaml"
+        else:
+            workload_dir_path = os.path.join(self.__dir_path, self.__workload_name)
+            data_filename = os.path.join('..', f'{self.__workload_name}_data_template.yaml')
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/common/template_operations/template_operations.py:102]: this one is a relative path, so it cannot be changed easily
it's used there:
> workload_data = yaml.load(render_yaml_file(workload_dir_path, data_filename

[the remark this answers is not in the document]
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/template_operations.py:103]: OK.

File: benchmark_runner/main/main.py (hunk 9 above)
Comment (RobertKrawitz) [reply benchmark_runner/main/main.py:162]: Agreed.  I hate seeing stack traces out of every error.

File: benchmark_runner/main/environment_variables.py [hunk 19]
```diff
@@ -13,6 +13,23 @@ def __init__(self):
...
+                with open(env) as f:
+                    for line in f.readlines():
+                        key, found , value = line.strip().partition("=")
+                        if not found:
+                            print("ERROR: invalid line in {env}: {line.strip()}")
+                            continue
+                        if key in os.environ: continue # prefer env to env file
+                        os.environ[key] = value
+
+                        print(f"{key} = {value}")
```
Comment (ebattat) [inline benchmark_runner/main/environment_variables.py:30]: Change print to logger.info, if it not necessary delete it at all

File: benchmark_runner/main/environment_variables.py [hunk 20]
```diff
@@ -88,11 +107,22 @@ def __init__(self):
         self._environment_variables_dict['date_key'] = datetime.datetime.now().strftime("%Y/%m/%d")
         self._environment_variables_dict['time_stamp_format'] = os.path.join(datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d-%H-%M-%S'))
         # Benchmark runner local run artifacts path with time stamp format
-        self._environment_variables_dict['run_artifacts_path'] = os.path.join(self._environment_variables_dict['run_artifacts'], f"{self._environment_variables_dict['workload'].replace('_', '-')}-{self._environment_variables_dict['time_stamp_format']}")
+
+        self._environment_variables_dict['run_artifacts_path'] = os.environ.get('RUN_ARTIFACTS_PATH')
+        if not self._environment_variables_dict['run_artifacts_path']:
+            self._environment_variables_dict['run_artifacts_path'] = os.path.join(self._environment_variables_dict['run_artifacts'], f"{self._environment_variables_dict['workload'].replace('_', '-')}-{self._environment_variables_dict['time_stamp_format']}")
+        print("Saving artifacts into", self._environment_variables_dict['run_artifacts_path'])
```
Comment (ebattat) [inline benchmark_runner/main/environment_variables.py:114]: Change print to logger.info, if it not necessary delete it at all

File: benchmark_runner/main/main.py (hunk 9 above)
Comment (ebattat) [reply benchmark_runner/main/main.py:162]: The right way is to raise a custom Exception

Comment (kpouget) [reply benchmark_runner/main/main.py:162]: @ebattat you can't ignore my comment and not give any source for what you ask for ....

https://docs.python.org/3/library/sys.html#sys.exit
```
sys.exit([arg])
    Raise a SystemExit exception, signaling an intention to exit the interpreter.
```
I don't see anything against using it from the `main()` function (nor anywhere else in fact).

Raising a _custom_ exception in the `main()` is wrong, we need a way to define the return code of the process, which is the standard Unix convention to define if the process succeeded / why it failed.
```
raise SystemExit(N)
```
is another alternative, also it's semantically identical to the current code.

**Comment by ebattat** [comment]
@kpouget, Did you rebase it ?

**Comment by kpouget** [comment]
> @kpouget, Did you rebase it ?

yes, rebased and resolved the conflict

File: benchmark_runner/workloads/workloads.py [hunk 21]
```diff
@@ -22,16 +22,38 @@ def run(self):
...
+            initialize_workload = lambda: _initialize_workload(self)
+        except AttributeError:
+            print(f"WARNING: {workload} module has no initialize_workload method. Using the default one.")
+            initialize_workload = self.initialize_workload
+
         try:
-            self.initialize_workload()
-            # kata use pod module - replace kata to pod
-            workload = self._workload.replace('kata', 'pod')
+            _finalize_workload = getattr(workload_module, "finalize_workload")
```
[the remark this answers is not in the document]
Comment (kpouget) [reply benchmark_runner/workloads/workloads.py:39]: updated, also added the report of benchmark failure to benchmark-operator

File: benchmark_runner/workloads/workloads_operations.py [hunk 22]
```diff
@@ -18,6 +18,7 @@
 
 
 class WorkloadsOperations:
+    oc = None
```
Comment (ebattat) [inline benchmark_runner/workloads/workloads_operations.py:21]: @kpouget, What the purpose of 'oc = None' parameter if you are not using it. can you explain ?
